chapters/part3/hyperparameter-optimization-with-optuna.md

```diff
@@ -0,0 +1,362 @@
+<a href="https://optuna.org">Optuna</a> is a library for hyperparameter optimization, which allow users to easily optimize hyperparameters automatically.
+Optuna provides sophisticated algorithms for searching parameters, such as [Tree-structured Parzen Estimator](https://papers.nips.cc/paper/4443-algorithms-for-hyper-parameter-optimization) and
+[CMA Evolution Strategy](https://arxiv.org/abs/1604.00772), as well as algorithms for pruning unpromising trials, such as [Hyperband](http://jmlr.org/papers/v18/16-558.html).
+
+</exercise>
+
+<exercise id="2" title="Building your model">
+
+This tutorial works on sentiment analysis, one kind of text classification.
+We use the IMDb review dataset, which contains 20,000 positive/negative reviews for training and 5,000 reviews for validating the performance of model.
+If you haven't read <a href="https://guide.allennlp.org/your-first-model#1">the tutorial for text classification</a>, that may be helpful.
+
+Below is a sample configuration of a CNN-based classifier.
+Note that this model has six hyperparameters: `embedding_dim`, `dropout`, `lr`, `max_filter_size`, `num_filters`, and `output_dim`.
+
+```json
+// imdb_baseline.jsonnet
+
+local batch_size = 64;
+local cuda_device = 0;
+local num_epochs = 15;
+local seed = 42;
+local train_data_path = 'https://s3-us-west-2.amazonaws.com/allennlp/datasets/imdb/train.jsonl';
+local validation_data_path = 'https://s3-us-west-2.amazonaws.com/allennlp/datasets/imdb/dev.jsonl';
+
+// hyperparameters
+local embedding_dim = 128;
+local dropout = 0.2;
+local lr = 0.1;
+local max_filter_size = 4;
+local num_filters = 128;
+local output_dim = 128;
+local ngram_filter_sizes = std.range(2, max_filter_size);
+
+{
+  numpy_seed: seed,
+  pytorch_seed: seed,
+  random_seed: seed,
+  dataset_reader: {
+    type: 'text_classification_json',
+    tokenizer: {
+      type: 'whitespace',
+    },
+    token_indexers: {
+      tokens: {
+        type: 'single_id',
+      },
+    },
+  },
+  datasets_for_vocab_creation: ['train'],
+  train_data_path: train_data_path,
+  validation_data_path: validation_data_path,
+  model: {
+    type: 'basic_classifier',
+    text_field_embedder: {
+      token_embedders: {
+        tokens: {
+          embedding_dim: embedding_dim,
+        },
+      },
+    },
+    seq2vec_encoder: {
+      type: 'cnn',
+      embedding_dim: embedding_dim,
+      ngram_filter_sizes: ngram_filter_sizes,
+      num_filters: num_filters,
+      output_dim: output_dim,
+    },
+    dropout: dropout,
+  },
+  data_loader: {
+    batch_size: batch_size,
+  },
+  trainer: {
+    cuda_device: cuda_device,
+    num_epochs: num_epochs,
+    optimizer: {
+      lr: lr,
+      type: 'sgd',
+    },
+    validation_metric: '+accuracy',
+  },
+}
+```
+
+These hyperparameters are selected based on the standard recommendations.
+Of course, we can train this model using AllenNLP CLI.
+I ran `allennlp train imdb_baseline.jsonnet` five times with random seeds.
```

Comment:
How about showing that you used different random seeds?

Reply:
f954dcc Thank you. :pray: